### PR TITLE
RDSE: serialization for py bindings

### LINKED ISSUE #600: Python encoders serialization
Hi everybody. I recently get this package working and I am using the python3 bindings.
Nevertheless, I have encountered some problems while trying to serialize encoders and encoder parameters.

Here is the code sample I have done : 

```
In [95]: params = RDSE_Parameters()

In [96]: params.size = 100

In [97]: params.sparsity = 0.1

In [98]: params.resolution = 0.1

In [99]: rdse = RDSE(params)

In [100]: with open("save", "wb") as f:
     ...:     pickle.dump(params, f)
     ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-100-6147ac3148f8> in <module>
      1 with open("save", "wb") as f:
----> 2     pickle.dump(params, f)
      3

TypeError: can't pickle htm.bindings.encoders.RDSE_Parameters objects

In [101]: with open("save", "wb") as f:
     ...:     pickle.dump(rdse, f)
     ...:
     ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-101-d8cd9ca6b44e> in <module>
      1 with open("save", "wb") as f:
----> 2     pickle.dump(rdse, f)
      3
      4

TypeError: can't pickle htm.bindings.encoders.RDSE objects
```

which is just creating a RDSE_parameters() object and then a RDSE() object, and trying to serialize them with pickle. This apparently does not work, although serialization works fine for spatial poolers and temporal memories. 
My question is then : will the pickle serialization of the encoders be available soon? Otherwise, what are the alternatives for serializing the rdse and rdse_parameters objects?

Thank you by advance for you answers. 

here is another sample of code where I try to serialize a date encoder. Since the date encoder is just an implementation of a rdse encoder, I run into the same error : 

```
In [102]: from htm.encoders.date import DateEncoder

In [103]: date_enc = DateEncoder(season=10)

In [104]: with open("save", "wb") as f:
     ...:     pickle.dump(date_enc, f)
     ...:
     ...:
     ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-104-13d5f28aa5c5> in <module>
      1 with open("save", "wb") as f:
----> 2     pickle.dump(date_enc, f)
      3
      4
      5

TypeError: can't pickle htm.bindings.encoders.ScalarEncoder objects
```

### PR DESCRIPTION
**Pull request by breznak**
TODO
- [x] py test for pickle 

Fixes #600

### CODE REVIEW

**Comment by breznak** [comment]
This is waiting for @leonardbcm for the tests, but if we don't hear from him, can be merged as is

**Comment by Leonardbcm** [comment]
Hi, thanks for the support again.

I tried serializing and deserializing rdse objects with pickle and it worked fine. 

I have written this following piece of code : 

`def testPickle(self):
        """
        The pickling is successfull if pickle serializes and de-serialize the
        RDSE object. 
        Moreover, the de-serialized object shall give the same SDR than the 
        original encoder given the same scalar value to encode.
        """        
        rdse_params = RDSE_Parameters()
        rdse_params.sparsity = 0.1
        rdse_params.size = 100
        rdse_params.resolution = 0.1
        rdse_params.seed = 1997

        rdse = RDSE(rdse_params)
        filename = "RDSE_testPickle"

        try:
            with open(filename, "wb") as f:
                pickle.dump(rdse, f)
        except:
            dump_success = False
        else:
            dump_success = True

        assert(dump_success)

        try:
            with open(filename, "rb") as f:
                rdse_loaded = pickle.load(f)
        except:
            read_success = False
        else:
            read_success = True

        assert(read_success)
        value_to_encode = 69003        
        SDR_original = rdse.encode(value_to_encode)
        SDR_loaded = rdse_loaded.encode(value_to_encode)

        assert(SDR_original == SDR_loaded)`

Which gives the following logs : 

`C:\Users\Leonard\Documents\htm.core>pytest bindings/py/tests/encoders/rdse_test.py
============================================================================================== test session starts ===============================================================================================
platform win32 -- Python 3.7.2, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- c:\users\leonard\appdata\local\programs\python\python37\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\Leonard\Documents\htm.core, inifile: setup.cfg
plugins: cov-2.7.1
collected 9 items

bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testAverageOverlap PASSED                                                                                                                               [ 11%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testConstructor PASSED                                                                                                                                  [ 22%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testDeterminism PASSED                                                                                                                                  [ 33%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testErrorChecks PASSED                                                                                                                                  [ 44%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testPickle FAILED                                                                                                                                       [ 55%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testRadiusResolution PASSED                                                                                                                             [ 66%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testRandomOverlap PASSED                                                                                                                                [ 77%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testSeed PASSED                                                                                                                                         [ 88%]
bindings/py/tests/encoders/rdse_test.py::RDSE_Test::testSparsityActiveBits PASSED                                                                                                                           [100%]

==================================================================================================== FAILURES ====================================================================================================
______________________________________________________________________________________________ RDSE_Test.testPickle ______________________________________________________________________________________________

self = <rdse_test.RDSE_Test testMethod=testPickle>

    def testPickle(self):
        """
        The pickling is successfull if pickle serializes and de-serialize the
        RDSE object.
        Moreover, the de-serialized object shall give the same SDR than the
        original encoder given the same scalar value to encode.
        """
        rdse_params = RDSE_Parameters()
        rdse_params.sparsity = 0.1
        rdse_params.size = 100
        rdse_params.resolution = 0.1
        rdse_params.seed = 1997

        rdse = RDSE(rdse_params)
        filename = "RDSE_testPickle"

        try:
            with open(filename, "wb") as f:
                pickle.dump(rdse, f)
        except:
            dump_success = False
        else:
            dump_success = True

        assert(dump_success)

        try:
            with open(filename, "rb") as f:
                rdse_loaded = pickle.load(f)
        except:
            read_success = False
        else:
            read_success = True

        assert(read_success)
        value_to_encode = 69003
        SDR_original = rdse.encode(value_to_encode)
>       SDR_loaded = rdse_loaded.encode(value_to_encode)
E       RuntimeError: CHECK FAILED: "output.size == size"

bindings\py\tests\encoders\rdse_test.py:295: RuntimeError
---------------------------------------------------------------------------------------------- Captured stderr call ----------------------------------------------------------------------------------------------
ERR:  CHECK FAILED: "output.size == size"  [c:\users\leonard\documents\htm.core\src\htm\encoders\randomdistributedscalarencoder.cpp line 94]
----------------------------------------------------------------- generated xml file: C:\Users\Leonard\Documents\htm.core\junit-test-results.xml -----------------------------------------------------------------
======================================================================================= 1 failed, 8 passed in 0.42 seconds =======================================================================================`

The loaded encoder does not seems to be able to encoder values anymore.
I will investigate this next Monday.

**Comment by breznak** [comment]
> I have written this following piece of code 

the test looks good, please add it to this PR.

**Comment by Leonardbcm** [comment]
Hi, while trying to push comited changes on "serialization_rdse', the permission is denied.

**Comment by breznak** [comment]
> while trying to push comited changes on "serialization_rdse', the permission is denied.

yes, you don't have push access to this repo. The work flow should be: 
- you fork this repo to your account
- develop the changes (you are here) 
- push the branch to your repo
- open a PR (your feature branch -> this repo) 
  - in this case, do not open against master here, but against serialization_rdse branch
- I'll then merge the changes. 

The problem is because I've started a branch from this repo, instead of a PR from your repo which would normally occur.

**Comment by breznak** [comment]
Back to this branch. 
@Leonardbcm are the tests passing for you? @brev had a similar problem with failing bindings on pickle in #603

**Comment by breznak** [comment]
Ping @Leonardbcm , please test the code

File: bindings/py/cpp_src/bindings/encoders/py_RDSE.cpp [hunk 1]
```diff
@@ -113,5 +115,37 @@ fields are filled in automatically.)");
...
+          },
+          [](py::bytes &s) {
+            std::stringstream ss( s.cast<std::string>() );
+            RDSE self;
+            self.load(ss);
+            return self;
+        }));
+
     }
 }
```
Comment (dkeeney) [inline bindings/py/cpp_src/bindings/encoders/py_RDSE.cpp]: Also add these bindings:
```
        py_RDSE.def("saveToFile", 
		 [](RDSE &self, const std::string& filename) {self.saveToFile(filename,SerializableFormat::BINARY); });  
				
        py_RDSE.def("loadFromFile",
		 [](RDSE &self, const std::string& filename) { return self.loadFromFile(filename,SerializableFormat::BINARY); });

File: bindings/py/tests/encoders/rdse_test.py [hunk 2]
```diff
@@ -255,6 +255,44 @@ def testSeed(self):
...
+            read_success = False
+        else:
+            read_success = True
+
+        assert(read_success)
+        value_to_encode = 69003        
+        SDR_original = rdse.encode(value_to_encode)
+        SDR_loaded = rdse_loaded.encode(value_to_encode)
+
+        assert(SDR_original == SDR_loaded)
```
Comment (dkeeney) [inline bindings/py/tests/encoders/rdse_test.py]: Also add a test for saveToFile() and loadFromFile()
See end of spatial_pooler_test.py  in branch sp_save.

**Comment by breznak** [comment]
Thanks @dkeeney , this is working with your fixes from #644

**Comment by breznak** [comment]
@Leonardbcm @dkeeney can you ACK the code, please? There's a remaining test, but I don't think I'll have time to write it in near future, so I'd like to merge the fixes as is, which makes the requested serialization for the encoders work.

**Comment by breznak** [comment]
Thank you! 

> I am working on a generic Encoder Region

this will be interesting!

**Pull request merged**